apps/cli/src/program.ts

```diff
@@ -0,0 +1,133 @@
+import { Command } from "commander";
+import * as path from "path";
+import { watch } from "chokidar";
+import * as fs from "fs/promises";
+import { parseMachinesFromFile } from "@xstate/machine-extractor";
+import {
+  doesTsTypesRequireUpdate,
+  FileEdit,
+  makeXStateUpdateEvent,
+  processFileEdits,
+  writeToTypegenFile,
+} from "@xstate/tools-shared";
+import { version } from "../package.json";
+
+export const program = new Command();
+
+program.version(version);
+
+const handleError = (uri: string, e: any) => {
+  if (e?.code === "BABEL_PARSER_SYNTAX_ERROR") {
+    console.log(`${uri} - syntax error, skipping`);
+  }
+  if (process.env.NODE_ENV === "test") {
+    console.log(e);
+  }
+};
+
+const writeToFiles = async (uriArray: string[]) => {
+  /**
+   * TODO - implement pretty readout
+   */
+  await Promise.all(
+    uriArray.map(async (uri) => {
+      try {
+        const fileContents = await fs.readFile(uri, "utf8");
+        const parseResult = parseMachinesFromFile(fileContents);
+
+        if (!parseResult.machines.length) {
+          return;
+        }
+
+        const event = makeXStateUpdateEvent(
+          uri,
+          parseResult.machines.map((machine) => ({
+            parseResult: machine,
+          })),
+        );
+
+        const fileEdits: FileEdit[] = [];
+
+        let machineIndex = 0;
+
+        for (const machine of parseResult.machines) {
+          if (machine.ast.definition?.tsTypes?.node) {
+            const { name } = path.parse(uri);
+            const requiresUpdate = doesTsTypesRequireUpdate({
+              fileText: fileContents,
+              machineIndex,
+              node: machine.ast.definition.tsTypes.node,
+              relativePath: name,
+            });
+
+            if (requiresUpdate) {
+              fileEdits.push({
+                start: machine.ast.definition.tsTypes.node.start!,
+                end: machine.ast.definition.tsTypes.node.end!,
+                newText: `{} as import("./${name}.typegen").Typegen${machineIndex}`,
+              });
+            }
+            machineIndex++;
+          }
+        }
+
+        if (fileEdits.length > 0) {
+          const newFile = processFileEdits(fileContents, fileEdits);
+
+          await fs.writeFile(uri, newFile);
+        }
+
+        await writeToTypegenFile({
+          filePath: uri,
+          event,
+        });
+        console.log(`${uri} - success`);
+      } catch (e) {
+        handleError(uri, e);
+      }
+    }),
+  );
+};
+
+export const runTypegen = (filesPattern: string, opts: { watch?: boolean }) => {
+  return new Promise<void>((resolve, reject) => {
```

Comment:
Returning a promise from something that is passed to the `.action` might have a special meaning there. This is a little bit iffy for me, especially given that only one specific branch here returns a promise. 

This could be restructured in a way that you'd export a function returning a promise, but one that wouldn't even deal with `chokidar`. So basically the inner `writeToFiles` could be renamed, promisified (if it's not already), and exported from here.

Reply:
Ah, i guess u are resolving the glob with chokidar, but we could simply just read all filenames in that dir and skip over that.

Reply:
Yeah I'm trying to test as much of the production code as possible. I found it hard to test the CLI itself (probably missing some knowledge about how to test commander apps).

> This is a little bit iffy for me, especially given that only one specific branch here returns a promise.

Ah, you're right. I'll have a think.

Reply:
> Yeah I'm trying to test as much of the production code as possible. I found it hard to test the CLI itself (probably missing some knowledge about how to test commander apps).

Hm, couldn't u `execSync` the compiled CLI? 🤔

Reply:
Yeah that's what I tried, will try again.